### PR TITLE
Update documentation links

### PR DESCRIPTION
**Pull request by tybug**
As I was looking at adding an automated check for outdated links, I found some broken links in our documentation. This pull fixes those, and updates most 301 permanent redirects, and some 302 Found redirects.

To run locally:

```
sphinx-build --builder linkcheck hypothesis-python/docs linkcheck
```

I've also written a sphinx extension that modifies the linkcheck builder to ignore intersphinx links. I think we can rely on those being accurate.

I was thinking we could run linkcheck on each pull in the future? `linkcheck` still picks up on a few false positive request timeouts or missing section headers though, so we'll need a way to manually whitelist false positives before turning such a job on.

### CODE REVIEW

File: hypothesis-python/docs/changes.rst [hunk 1]
```diff
@@ -1930,7 +1930,7 @@ This release further improves our ``.patch``-file support from
 :func:`~hypothesis.strategies.data` (and don't support
 :obj:`@example() <hypothesis.example>`\ ), and various broken edge-cases.
 
-Because :pypi:`libCST` has released version 1.0 which uses the native parser
+Because :pypi:`libCST <libcst>` has released version 1.0 which uses the native parser
```
Comment (tybug) [inline hypothesis-python/docs/changes.rst:1933]: changes in this vein are to silence 302 redirect warnings from linkcheck, and make combing through the remaining output easier

File: hypothesis-python/docs/endorsements.rst [hunk 2]
```diff
@@ -27,7 +27,7 @@ has caught or prevented far more bugs - than our old style of
 example-based testing.
 
 ---------------------------------------------------------------------------------------
-Kristian Glass - Director of Technology at `LaterPay GmbH <https://www.laterpay.net/>`_
+Kristian Glass - Director of Technology at LaterPay GmbH
```
Comment (tybug) [inline hypothesis-python/docs/endorsements.rst:30]: Website 404s, and I figured no link was better than a broken one?

File: hypothesis-python/docs/changes.rst [hunk 3]
```diff
@@ -402,7 +402,7 @@ metadata and messages in our :doc:`observability output <observability>`
 --------------------
 
 Support ``__default__`` field of :obj:`~python:typing.TypeVar`
-and support the same from :pypi:`typing_extensions`
+and support the same from :pypi:`typing_extensions <typing-extensions>`
```
Comment (Zac-HD) [inline hypothesis-python/docs/changes.rst:405]:
```suggestion
and support the same from :pypi:`typing-extensions`
```
I'd be inclined to just spell this with the hyphen, and similarly below

**Review by Zac-HD — APPROVED** [review]
Looks great!  Merge with or without the notes below 🙂 

> I was thinking we could run linkcheck on each pull in the future? linkcheck still picks up on a few false positive request timeouts or missing section headers though, so we'll need a way to manually whitelist false positives before turning such a job on.

I'm -1 on this in CI; flaky jobs are generally terrible.  We could set it up as a ~weekly cronjob though, which would make it easy to check in occasionally just by clicking through.

**Comment by tybug** [comment]
The motivation for a ci job was not requiring me to remember to check weekly for broken links, but I definitely see the point about flakiness. We'll shelve that for now then.

**Pull request merged**